android/src/main/java/com/oblador/keychain/KeychainModule.java

```diff
@@ -97,59 +117,96 @@ public void setGenericPasswordForOptions(String service, String username, String
     }
 
     @ReactMethod
-    public void getGenericPasswordForOptions(String service, Promise promise) {
+    public void getGenericPasswordForOptions(String service, final Promise promise) {
+        final String serviceOrDefault = getDefaultServiceIfNull(service);
+        CipherStorage cipherStorage = null;
         try {
-            service = getDefaultServiceIfNull(service);
-
-            CipherStorage currentCipherStorage = getCipherStorageForCurrentAPILevel();
-
-            ResultSet resultSet = prefsStorage.getEncryptedEntry(service);
+            ResultSet resultSet = prefsStorage.getEncryptedEntry(serviceOrDefault);
             if (resultSet == null) {
-                Log.e(KEYCHAIN_MODULE, "No entry found for service: " + service);
+                Log.e(KEYCHAIN_MODULE, "No entry found for service: " + serviceOrDefault);
                 promise.resolve(false);
                 return;
             }
 
-            final DecryptionResult decryptionResult = decryptCredentials(service, currentCipherStorage, resultSet);
-
-            WritableMap credentials = Arguments.createMap();
-
-            credentials.putString("service", service);
-            credentials.putString("username", decryptionResult.username);
-            credentials.putString("password", decryptionResult.password);
+            // Android < M will throw an exception as biometry is not supported.
+            CipherStorage biometryCipherStorage = null;
+            try {
+                biometryCipherStorage = getCipherStorageForCurrentAPILevel(true);
+            } catch(Exception e) { }
+            final CipherStorage nonBiometryCipherStorage = getCipherStorageForCurrentAPILevel(false);
+            if (biometryCipherStorage != null && resultSet.cipherStorageName.equals(biometryCipherStorage.getCipherStorageName())) {
+                cipherStorage = biometryCipherStorage;
+            } else if (nonBiometryCipherStorage != null && resultSet.cipherStorageName.equals(nonBiometryCipherStorage.getCipherStorageName())) {
+                cipherStorage = nonBiometryCipherStorage;
```

Comment:
there is `cipherStorage`, `biometryCipherStorage`, `nonBiometryCipherStorage`... If I understand this correctly, you're trying to find out which is the preferred storage given the params passed from JS.

Would it make sense to have `preferredCipherStorage` and `fallbackCipherStorage` variable names, and extract this code block (lines 131 to 140) into a separate class with a understandable name (eg. CipherStorageFinder.determinePreferredAndFallbackCipherStorages()` - just an idea)? This is easy to do via refactor -> extract (albeit the resulting java code is meh, but that's java)